src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+    *         in this list, so it is not necessary to perform these computations on the returned
+    *         sequence.
+    */
+  @tailrec
+  def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
+                  origin: (Double, Double, Double)): Seq[Block] = {
+    val reconstructedBlocks = (processorBlocks map { block =>
+      val currentBlock = Block(origin, processorBlocks.head.updateFaces(origin))
+      val updatedBlock = Block(origin, block.updateFaces(origin))
+      if (!currentBlock.approximateEquals(updatedBlock)) {
+        val sharedProcFaces = compareProcessorBlocks(currentBlock, updatedBlock)
+        if (sharedProcFaces.nonEmpty) {
+          val currentFaces = currentBlock.faces.diff(sharedProcFaces)
+          val updatedFaces = updatedBlock.faces.diff(sharedProcFaces)
+          val allFaces = currentFaces ++ updatedFaces
+          // Check for any actual shared faces between blocks - if these exists blocks
+          // should not be merged since there is a real joint seperating the blocks
+          val nonSharedFaces =
+            allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
+              if (!unique.exists(compareSharedFaces(current, _))) {
+                current +: unique
+              } else {
+                unique
+              }
+            }
+          if (allFaces.diff(nonSharedFaces).isEmpty) {
+            Block(origin, nonSharedFaces)
+          }
         }
-        if (right.nonEmpty) {
-          queue.enqueue(right)
+      }
+    }).collect { case blockType: Block => blockType }
+
+    // Remove redundant faces and remove duplicates
+    val nonRedundantBlocks = reconstructedBlocks map { case block @ Block(center, _) =>
+      Block(center, block.nonRedundantFaces)
+    }
+    val joinedBlocks = 
+      nonRedundantBlocks.foldLeft(Seq.empty[Block]) { (unique, current) =>
+        if (!unique.exists(current.approximateEquals(_))) {
+          current +: unique
+        } else {
+          unique
         }
-        median +: loadBalanceIndicesImpl(queue)
       }
+
+    // Divide blocks into groups with and without processor joints remaining
+    val (remainingBlocks, completedBlocks) = joinedBlocks.partition { block =>
+      block.faces.exists(_.processorJoint)
     }
 
-    def removeMedian[A](s: Seq[A]): (A, Seq[A], Seq[A]) = {
-      val medianIndex = s.length / 2
-      (s(medianIndex), s.take(medianIndex), s.drop(medianIndex + 1))
+    if (remainingBlocks.nonEmpty) {
+      // Merged blocks still contain some processor joints
+      mergeBlocks(remainingBlocks ++ processorBlocks.tail, completedBlocks ++ mergedBlocks, origin)
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
```

Comment:
This duplicate checking needs to stay -  when `remainingBlocks` gets prepended to `processorBlocks.tail` there is the possibility that the same blocks can be generated through different merging orders.